File: packages/ios/Sources/FernlinkSDK/transport/TransportMessageRouter.swift
import Foundation

final class TransportMessageRouter {

    private let transport:  FernlinkTransport
    private let keypair:    FernlinkKeypair
    private let rpc:        SolanaRpc
    private let proofStore: ProofStore

    private let routerLock             = NSLock()
    private var _currentTxSig:         String              = ""
    private var _collectedProofs:      [VerificationProof] = []
    private var _seenVerifierKeys:     Set<Data>           = []
    private var _originatedRequestIds: Set<String>         = []

    var externalRequestSender: ((Data) -> Void)?
    var externalProofSender:   ((Data) -> Void)?
    var onMeshEvent:           ((String) -> Void)?

    init(
        transport:   FernlinkTransport,
        keypair:     FernlinkKeypair,
        rpcEndpoint: String,
        proofStore:  ProofStore
    ) {
        self.transport  = transport
        self.keypair    = keypair
        self.rpc        = SolanaRpc(rpcEndpoint)
        self.proofStore = proofStore
    }

    func setExternalForwarders(
        requestSender: @escaping (Data) -> Void,
        proofSender:   @escaping (Data) -> Void
    ) {
        externalRequestSender = requestSender
        externalProofSender   = proofSender
    }

    func start() {
        transport.onIncomingRequest = { [weak self] data in self?.handleIncomingRequest(data) }
        transport.onIncomingProof   = { [weak self] data in self?.handleIncomingProof(data) }
    }

    func clearProofs() {
        routerLock.withLock {
            _collectedProofs       = []
            _seenVerifierKeys      = []
            _originatedRequestIds  = []
            _currentTxSig          = ""
        }
    }

    func collectedProofsList() -> [VerificationProof] {
        routerLock.withLock { _collectedProofs }
    }

    func broadcastRequest(txSignature: String, commitment: String = "confirmed", ttl: Int = 8) {
        let requestId = UUID().uuidString
        routerLock.withLock {
            _currentTxSig = txSignature
            _originatedRequestIds.insert(requestId)
        }
        if transport.connectedPeerCount == 0 {
            proofStore.enqueue(.init(txSignature: txSignature, commitment: commitment, ttl: ttl))
            return
        }
        sendRequest(txSignature: txSignature, commitment: commitment, ttl: ttl, requestId: requestId)
    }

    // Relay a request from another transport to this transport's peers — no local RPC.
    func injectRequest(_ data: Data) {
        transport.sendRequest(data)
    }

    // Accept a proof from another transport: verify, dedup, collect, forward to local peers.
    // Does NOT call externalProofSender to prevent inter-transport bouncing.
    func injectProof(_ data: Data) {
        guard let proof = try? JSONDecoder().decode(VerificationProof.self, from: data),
              verifyProof(proof)
        else { return }
        let added: Bool = routerLock.withLock {
            // proof.txSignature holds at most 64 chars (wire format truncates to 64 bytes).
            let txKey = String(_currentTxSig.prefix(64))
            guard !txKey.isEmpty, proof.txSignature == txKey else { return false }
            guard _seenVerifierKeys.insert(proof.verifierPublicKey).inserted else { return false }
            _collectedProofs.append(proof)
            return true
        }
        if added { transport.sendProof(data) }
    }

    // MARK: - Private

    private func handleIncomingRequest(_ data: Data) {
        Task {
            guard let json = try? JSONDecoder().decode(VerificationRequest.self, from: data) else { return }
            if let reqId = json.requestId {
                let isOwn = routerLock.withLock { _originatedRequestIds.contains(reqId) }
                if isOwn { return }
            }
            onMeshEvent?("REQUEST_IN:\(json.txSignature.prefix(20))")
            do {
                onMeshEvent?("RPC_QUERYING")
                let status = try await rpc.getSignatureStatus(json.txSignature)
                let proof  = try keypair.signProof(
                    txSignature: json.txSignature,
                    status:      status.status,
                    slot:        status.slot,
                    blockTime:   status.blockTime,
                    errorCode:   0
                )
                if let proofData = try? JSONEncoder().encode(proof) {
                    transport.sendProof(proofData)
                    let statusStr: String
                    switch proof.status {
                    case .confirmed: statusStr = "Confirmed"
                    case .failed:    statusStr = "Failed"
                    case .unknown:   statusStr = "Unknown"
                    }
                    onMeshEvent?("PROOF_SENT:\(statusStr):\(proof.slot)")
                }
            } catch {
                onMeshEvent?("RPC_FAIL")
                if json.ttl > 0 {
                    let forwarded = VerificationRequest(
                        txSignature: json.txSignature,
                        commitment:  json.commitment,
                        ttl:         json.ttl - 1,
                        requestId:   json.requestId
                    )
                    if let fwdData = try? JSONEncoder().encode(forwarded) {
                        transport.sendRequest(fwdData)
                        externalRequestSender?(fwdData)
                        onMeshEvent?("FORWARDING:\(json.ttl - 1)")
                    }
                }
            }
        }
    }

    private func handleIncomingProof(_ data: Data) {
        guard let proof = try? JSONDecoder().decode(VerificationProof.self, from: data),
              verifyProof(proof)
        else { return }
        let (added, count): (Bool, Int) = routerLock.withLock {
            let txKey = String(_currentTxSig.prefix(64))
            guard !txKey.isEmpty, proof.txSignature == txKey else { return (false, 0) }
            guard _seenVerifierKeys.insert(proof.verifierPublicKey).inserted else { return (false, 0) }
            _collectedProofs.append(proof)
            return (true, _collectedProofs.count)
        }
        if added {
            let pubKey = proof.verifierPublicKey.map { String(format: "%02x", $0) }.joined()
            onMeshEvent?("PROOF_RECV:\(count):\(pubKey)")
            transport.sendProof(data)
            externalProofSender?(data)
        }
    }

    private func sendRequest(txSignature: String, commitment: String, ttl: Int, requestId: String? = nil) {
        let req = VerificationRequest(txSignature: txSignature, commitment: commitment, ttl: ttl, requestId: requestId)
        guard let data = try? JSONEncoder().encode(req) else { return }
        transport.sendRequest(data)
    }
}

File: packages/android/fernlink-sdk/src/main/kotlin/xyz/fernlink/sdk/transport/TransportMessageRouter.kt
package xyz.fernlink.sdk.transport

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import org.json.JSONObject
import xyz.fernlink.sdk.FernlinkJni
import xyz.fernlink.sdk.SolanaRpc
import xyz.fernlink.sdk.TxStatus
import xyz.fernlink.sdk.ble.ProofStore
import java.util.Collections
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue

internal class TransportMessageRouter(
    private val incomingRequests: SharedFlow<ByteArray>,
    private val incomingProofs: SharedFlow<ByteArray>,
    private val sendProof: (ByteArray) -> Unit,
    private val sendRequest: (ByteArray) -> Unit,
    private val connectedPeerCount: () -> Int,
    private val proofStore: ProofStore,
    private val keypairSeed: ByteArray,
    private val rpcEndpoint: String,
    private val scope: CoroutineScope,
    context: Context? = null,
) {
    private val rpc = SolanaRpc(rpcEndpoint, context)
    private val collectedProofsList = ConcurrentLinkedQueue<String>()
    val collectedProofs: List<String> get() = collectedProofsList.toList()

    private val originatedRequestIds: MutableSet<String> =
        Collections.newSetFromMap(ConcurrentHashMap())

    private val seenVerifierKeys: MutableSet<String> =
        Collections.newSetFromMap(ConcurrentHashMap())

    @Volatile private var currentTxSig: String = ""

    private var externalRequestSender: ((ByteArray) -> Unit)? = null
    private var externalProofSender: ((ByteArray) -> Unit)? = null

    fun setExternalForwarders(requestSender: (ByteArray) -> Unit, proofSender: (ByteArray) -> Unit) {
        externalRequestSender = requestSender
        externalProofSender = proofSender
    }

    fun injectRequest(payload: ByteArray) {
        sendRequest(payload)
    }

    fun injectProof(payload: ByteArray) {
        scope.launch {
            val json = String(payload, Charsets.UTF_8)
            if (FernlinkJni.verifyProof(json) && proofMatchesCurrentRound(json)) {
                val pubKeyHex = extractPubKeyHex(json)
                if (seenVerifierKeys.add(pubKeyHex)) {
                    collectedProofsList.add(json)
                    sendProof(payload)
                    externalProofSender?.invoke(payload)
                }
            }
        }
    }

    fun clearProofs() {
        collectedProofsList.clear()
        originatedRequestIds.clear()
        seenVerifierKeys.clear()
        currentTxSig = ""
    }

    fun start() {
        incomingRequests
            .onEach { payload -> scope.launch { handleIncomingRequest(payload) } }
            .launchIn(scope)

        incomingProofs
            .onEach { payload ->
                val json = String(payload, Charsets.UTF_8)
                if (FernlinkJni.verifyProof(json)) {
                    if (!proofMatchesCurrentRound(json)) return@onEach
                    val pubKeyHex = extractPubKeyHex(json)
                    if (seenVerifierKeys.add(pubKeyHex)) {
                        collectedProofsList.add(json)
                        sendProof(payload)
                        externalProofSender?.invoke(payload)
                    }
                }
            }
            .launchIn(scope)
    }

    private fun extractPubKeyHex(json: String): String = runCatching {
        val arr = JSONObject(json).getJSONArray("verifier_pubkey")
        buildString {
            for (i in 0 until minOf(arr.length(), 32))
                append("%02x".format(arr.getInt(i) and 0xFF))
        }
    }.getOrDefault("")

    private fun proofMatchesCurrentRound(json: String): Boolean {
        if (currentTxSig.isEmpty()) return false
        val proofTxSig = runCatching {
            val arr = JSONObject(json).getJSONArray("tx_signature")
            val bytes = ByteArray(minOf(arr.length(), 64)) { arr.getInt(it).and(0xFF).toByte() }
            String(bytes, Charsets.UTF_8).trimEnd('\u0000')
        }.getOrDefault("")
        return proofTxSig == currentTxSig.take(64)
    }

    private suspend fun handleIncomingRequest(payload: ByteArray) {
        runCatching {
            val json       = JSONObject(String(payload, Charsets.UTF_8))
            val txSig      = json.getString("txSignature")
            val commitment = json.optString("commitment", "confirmed")
            val ttl        = json.optInt("ttl", 0).coerceIn(0, 8)
            val requestId  = json.optString("requestId", "")

            if (requestId.isNotEmpty() && originatedRequestIds.contains(requestId)) return

            try {
                val status = rpc.getSignatureStatus(txSig)
                val statusByte: Byte = when (status.status) {
                    TxStatus.CONFIRMED -> 0
                    TxStatus.FAILED    -> 1
                    TxStatus.UNKNOWN   -> 2
                }
                val proofJson = FernlinkJni.signProof(
                    keypairSeed = keypairSeed,
                    txSignature = txSig,
                    statusByte  = statusByte,
                    slot        = status.slot,
                    blockTime   = status.blockTime,
                    errorCode   = 0,
                ) ?: return
                sendProof(proofJson.toByteArray(Charsets.UTF_8))
            } catch (_: Exception) {
                if (ttl > 0) {
                    val forwarded = JSONObject().apply {
                        put("txSignature", txSig)
                        put("commitment",  commitment)
                        put("ttl",         ttl - 1)
                        if (requestId.isNotEmpty()) put("requestId", requestId)
                    }.toString().toByteArray(Charsets.UTF_8)
                    sendRequest(forwarded)
                    externalRequestSender?.invoke(forwarded)
                }
            }
        }
    }

    fun broadcastRequest(txSignature: String, commitment: String = "confirmed", ttl: Int = 8) {
        currentTxSig = txSignature
        val requestId = UUID.randomUUID().toString()
        originatedRequestIds.add(requestId)
        if (connectedPeerCount() == 0) {
            proofStore.enqueue(ProofStore.PendingRequest(txSignature, commitment, ttl))
            return
        }
        val payload = JSONObject().apply {
            put("txSignature", txSignature)
            put("commitment",  commitment)
            put("ttl",         ttl)
            put("requestId",   requestId)
        }.toString().toByteArray(Charsets.UTF_8)
        sendRequest(payload)
    }

    fun collectConsensusJson(minProofs: Int): String? {
        val proofs = collectedProofs
        if (proofs.isEmpty()) return null

        val first = runCatching { JSONObject(proofs.first()) }.getOrNull()
        val statusStr = when (first?.optString("status", "Unknown")?.lowercase()) {
            "confirmed" -> "confirmed"
            "failed"    -> "failed"
            else        -> "unknown"
        }
        return JSONObject().apply {
            put("settled",    proofs.size >= minProofs)
            put("status",     statusStr)
            put("slot",       first?.optLong("slot", 0) ?: 0L)
            put("blockTime",  first?.optLong("block_time", 0) ?: 0L)
            put("proofCount", proofs.size)
        }.toString()
    }
}
